sessionagent/response.go

```diff
@@ -0,0 +1,168 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015-2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package sessionagent
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/snapcore/snapd/logger"
+)
+
+// ResponseType is the response type
+type ResponseType string
+
+// "there are three standard return types: Standard return value,
+// Background operation, Error", each returning a JSON object with the
+// following "type" field:
+const (
+	ResponseTypeSync  ResponseType = "sync"
+	ResponseTypeAsync ResponseType = "async"
+	ResponseTypeError ResponseType = "error"
+)
+
+// Response knows how to serve itself, and how to find itself
+type Response interface {
+	ServeHTTP(w http.ResponseWriter, r *http.Request)
+}
+
+type resp struct {
+	Status int          `json:"status-code"`
+	Type   ResponseType `json:"type"`
+	Result interface{}  `json:"result,omitempty"`
+}
+
+type respJSON struct {
+	Type       ResponseType `json:"type"`
```

Comment:
I have mixed feelings here,  the consistency is good, I wouldn't design this at all this way though if I started fresh.  @chipaca should we try to redesign this how we would do a v3 of the general API or not worth the confusion?

Reply:
I'm still thinking about this

Reply:
I'd be happy to review the overall API design, but does that need to be tied to this initial PR?  At present, the only API endpoint is so trivial I don't think we'd need to worry about third parties relying on it.

This PR is already quite large, so I've been avoiding building anything more on top of it while it is in flux (e.g. exit on idle, connecting to the session bus, etc).